### PR TITLE
Issue 1363: Initial clippy lint testing

### PR DESCRIPTION
**Pull request by ozghimire**
Initial draft for [issue 1363](https://github.com/unicode-org/icu4x/issues/1363).

**Observation** : These lints are marked as unknown lints while running `cargo build`

```
warning: unknown lint: `indexing_slicing`
  --> components/calendar/src/lib.rs:19:9
   |
19 | #[allow(indexing_slicing, expect_used, panic)]
   |         ^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unknown_lints)]` on by default

warning: unknown lint: `expect_used`
  --> components/calendar/src/lib.rs:19:27
   |
19 | #[allow(indexing_slicing, expect_used, panic)]
   |                           ^^^^^^^^^^^

warning: unknown lint: `panic`
  --> components/calendar/src/lib.rs:19:40
   |
19 | #[allow(indexing_slicing, expect_used, panic)]
   |                                        ^^^^^```
```
^^^
Solved. I was missing clippy namespace around them like `clippy::panic`

### CODE REVIEW

**Comment by ozghimire** [comment]
I added few rules to components/calendar's, does this look like the right approach ?

**Comment by sffc** [comment]
Thanks for getting started!  Initial comments:

1. Ideally any `#[allow(...)]` overrides will annotate individual lines of code, not whole modules.
2. To do this incrementally, you could start by adding `#![deny(...)]` in the *lib.rs* file of a crate of your choice, and see if you can get that to work before you do it to all the other crates.

**Comment by ozghimire** [comment]
I was able to configure `#![deny(...)]` in lib.rs for component/calendar. The config applied for all the files inside calendar and needed multiple `#[allow()]` for it to pass `cargo clippy`. 

This strategy seems to work and will apply to all the component items.

File: components/calendar/src/iso.rs
Comment (sffc) [inline components/calendar/src/iso.rs:397]: "allow" lints should be at the line level (not the function level), and they should be accompanied by a comment explaining why it is safe to do the unwrap/index.

**Review by sffc — CHANGES_REQUESTED** [review]

File: components/calendar/src/iso.rs
Comment (ozghimire) [reply components/calendar/src/iso.rs:397]: These allows are needed as per the addition of deny clippy list on lib.rs . The strategies on how to handle these should change as part of that, we can try to create issues to fix these and reference on the lines as well. The reason behind that is not having full understanding of the code and it makes it harder to complete this specific issue.

Comment (ozghimire) [reply components/calendar/src/iso.rs:397]: Observation : Adding these attributes are experimental on expressions
`error[E0658]: attributes on expressions are experimental`

File: components/calendar/src/lib.rs
Comment (gregtatum) [inline components/calendar/src/lib.rs:4]: Praise: This approach looks correct to me.

**Comment by ozghimire** [comment]
@gregtatum  Updated as pre the request. And about the panic, if enabled it fails on `provider.rs line 31`. Tried adding inline allow but didn't responded to it.

File: components/calendar/src/japanese.rs
Comment (gregtatum) [inline components/calendar/src/japanese.rs:161]: suggestion: I think these errors are coming from the `tinystr!` macro implementation. I would suggest adding this suppression in the `utils/tiny_str` crate.

File: components/calendar/src/provider.rs
Comment (gregtatum) [inline components/calendar/src/provider.rs:31]: My guess is that there is an `unwrap` in the proc_macro `icu_provider::data_struct`

Comment (gregtatum) [reply components/calendar/src/provider.rs:31]:
```diff
diff --git a/components/calendar/src/lib.rs b/components/calendar/src/lib.rs
index df3c2aa29..2ee4cbe15 100644
--- a/components/calendar/src/lib.rs
+++ b/components/calendar/src/lib.rs
@@ -6,7 +6,7 @@
     clippy::unwrap_used,
     clippy::expect_used,
     // If enabled it fails on provider.rs line 31
-    //clippy::panic
+    clippy::panic
 )]
 #![cfg_attr(not(any(test, feature = "std")), no_std)]
 
diff --git a/provider/macros/src/lib.rs b/provider/macros/src/lib.rs
index ce71709d1..caca75fb3 100644
--- a/provider/macros/src/lib.rs
+++ b/provider/macros/src/lib.rs
@@ -111,6 +111,7 @@ fn data_struct_impl(attr: AttributeArgs, item: ItemStruct) -> TokenStream2 {
                     }
                 ));
             }
+            #[allow(clippy::panic)]
             _ => {
                 panic!("Invalid attribute to #[data_struct]")
             }
```

I'm still getting the error locally, but the suppressions need to be added inside the macro. Macros expand the code in place, so clippy rule is being applied to the expanded code, but it's being reported at where the proc_macro is defined.

**Review by gregtatum — CHANGES_REQUESTED** [review]
Marking as request changes as this is still a draft PR, and I have answered the questions.

**Comment by ozghimire** [comment]
Observation: 
- Added  clippy rules in `utils/tiny_str crate` but still needed these rules explicitly in `japanese.rs`.
- Added clippy rules in `provider/macros/src/lib.rs` still same issue on `provider.rs` needed to add outward declaration on the top of  the file to resolve the error.

File: components/calendar/src/japanese.rs
Comment (sffc) [reply components/calendar/src/japanese.rs:161]: +1. It looks like a large fraction of these come from the `tinystr!` macro.

File: components/calendar/src/iso.rs
Comment (sffc) [inline components/calendar/src/iso.rs:400]: Suggestion: For cases like this, please take the time to investigate the safety of the unwrap. In this case, I see that this is a private function with 3 call sites. Try to structure the code such that the clippy exception is closer to the call site where it is quickly obvious that there is no risk of panic.

**Comment by sffc** [comment]
@Manishearth Can you take a look at the issue where enabling the clippy unwrap_used lint triggers on instances of the `tinystr!()` macro and what we can do to avoid that?

**Comment by Manishearth** [comment]
@sffc the fix upstream is for these lints to potentially ignore stuff inside `const` contexts. but i'm unsure if people would want this in clippy (worth filing an issue!)

the fix here is to add the allow attribute to the const generated by the tinystr macro. i think we should just do that

**Comment by sffc** [comment]
> the fix here is to add the allow attribute to the const generated by the tinystr macro. i think we should just do that

What I definitely want to avoid is someone using `tinystr!()` and then being required to add a `#[allow(clippy::unwrap_used)]` to annotate it every time. Clippy annotations should be required very sparingly. If they are used everywhere, they become harder to review and they lose their potency.

> @sffc the fix upstream is for these lints to potentially ignore stuff inside const contexts. but i'm unsure if people would want this in clippy (worth filing an issue!)

If the const context is at fault, then we could wrap the contents of the macro in a panicky function, and put the suppression inside the panicky function.

**Comment by Manishearth** [comment]
> What I definitely want to avoid is someone using `tinystr!()` and then being required to add a `#[allow(clippy::unwrap_used)]` to annotate it every time. Clippy annotations should be required very sparingly. If they are used everywhere, they become harder to review and they lose their potency.
> 


Right, I don't want that either: I'm suggesting that we add a _single_ annotation to the _body_ of `tinystr!()` :smile:

**Comment by sffc** [comment]
> Right, I don't want that either: I'm suggesting that we add a single annotation to the body of tinystr!() smile

That's my preference as well. I was under the impression that this was already tried and it didn't work. But it does work:

https://github.com/unicode-org/icu4x/pull/1677

**Comment by sffc** [comment]
@ozghimire Can you merge in #1677 and see if you can get rid of the extra annotations then?

**Comment by ozghimire** [comment]
@sffc  Merged and pushed up with changes. Provider macro is interesting, even after inline allow for panic clippy it still needed file wide allow on `canander/src/provider.rs`

**Comment by sffc** [comment]
@ozghimire There seems to be an issue with your PR; the diff when merged into main should be only your changes, not changes from other people's PRs. This is most likely something you can fix in `git`. If you need help with `git`, there are many resources you can find with a quick Google search for something like "git tutorials".

**Comment by ozghimire** [comment]
@sffc @gregtatum This changes are updated in the entire codebase and should ressolve https://github.com/unicode-org/icu4x/issues/1363

File: components/calendar/src/iso.rs
Comment (gregtatum) [inline components/calendar/src/iso.rs:398]: issue: I think this issue number #1688 is incorrect.

File: utils/tinystr/src/ascii.rs
Comment (gregtatum) [inline utils/tinystr/src/ascii.rs:44]: issue: This is work from #1677.

suggestion: This is already merged into main, and you should be able to merge in the latest, per Shane's suggestion.

**Review by gregtatum — CHANGES_REQUESTED** [review]
Thanks, this is getting really close. There are still 2 issues that need resolving:

* Updating the issue number to point to the correct follow-up issue.
* Fixing the merge.

It would be nice to get feedback from @sffc that the results of the clippy lints are correct, and the suppressions are used as expected.

**Comment by ozghimire** [comment]
@gregtatum The requested changes have been completed and pushed up.

File: ffi/diplomat/src/lib.rs
Comment (sffc) [inline ffi/diplomat/src/lib.rs:12]: Issue: This is not correct. The `feature(alloc_error_handler)` should be enabled in some cases, but `#![no_std]` and `#![allow(clippy:upper_case_acronyms)]` should always be enabled. Please revert this back to what it was before and add the new clippy exclusions as a standalone cfg_attr.

File: provider/cldr/src/lib.rs
Comment (sffc) [inline provider/cldr/src/lib.rs:24]: Request: For now, exclude this particular crate (`icu_provider_cldr`) from the change. This crate is being tracked separately in #574

File: components/calendar/src/iso.rs
Comment (sffc) [inline components/calendar/src/iso.rs:398]: Praise: I like having an issue number `#1668` because we can close the issue once all of the TODOs in source code are gone.

**Review by sffc — COMMENTED** [review]
Thank you for all your work! Two more comments and then I think we can merge.

**Comment by ozghimire** [comment]
@sffc @gregtatum  The changes requested are pushed up and conflicts have been resolved.

File: ffi/diplomat/src/lib.rs
Comment (sffc) [reply ffi/diplomat/src/lib.rs:12]: Issue: This is still not correct. You should have

```
#![no_std]
#![allow(clippy::upper_case_acronyms)]
#![cfg_attr(
    not(test),
    deny(
        clippy::indexing_slicing,
        clippy::unwrap_used,
        clippy::expect_used,
        clippy::panic
    )
)]
```

File: provider/cldr/src/lib.rs
Comment (sffc) [reply provider/cldr/src/lib.rs]: Request: Please also undo the rest of the changes to the files in this crate.

Comment (sffc) [reply provider/cldr/src/lib.rs]: Actually, never mind. What you did is fine.

**Review by sffc — COMMENTED** [review]
Please fix up the diplomat lib.rs and then we can merge; thanks!

**Review by sffc — APPROVED** [review]
LGTM pending CI coming back green

**Review by gregtatum — APPROVED** [review]

**Comment by gregtatum** [comment]
Looks like there are still a few CI issues that need fixing. `cargo fmt` will format the code.

**Comment by sffc** [comment]
> @gregtatum @sffc Can anyone approve the CI pipeline run ?

Yes, approved. Sorry for the delay here; once you land your first PR then they should run automatically.

**Comment by ozghimire** [comment]
@sffc  These changes should get the CI process to green. Let's run them again.

**Comment by sffc** [comment]
There is a conflict; can you resolve the conflict first?

**Comment by ozghimire** [comment]
@sffc  @gregtatum  Issue #1363 should be resolved by PR after merge.

**Review by sffc — APPROVED** [review]

**Pull request merged**
